### PR TITLE
Fix support for static provisioning

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

This PR fixes support for managed resources that do not reference a class. Controllers built against crossplane-runtime will currently panic if they encounter a managed resource without a class reference. Any dynamically provisioned managed resource will have a class reference, but this breaks the static provisioning workflow in which a managed resource is explicitly provisioned, then later claimed.

The PR also fixes watch predicates for the static binding flow We currently support dynamic provisioning in the resource claim reconciler by using a watch predicate that allows either managed resources that directly reference a non-portable resource class of a given kind, or resource claims that reference a non-portable resource class of a given kind indirectly via a portable resource class.

To support static provisioning (i.e. explicitly claiming an existing managed resource) we must also allow resource claims that explicitly reference a managed resource. Writing one predicate to do all of this was getting cumbersome, so I have refactored the predicate interface a little.

This PR will require additional PRs to update Crossplane core, and each stack.

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

**Comment by negz** [comment]
I've tested that both the static and dynamic provisioning flows work by using a build of https://github.com/crossplaneio/stack-gcp/pull/24 to statically and then dynamically provision a `CloudMemorystoreInstance` and bind it to a `RedisCluster`.

File: pkg/resource/predicates.go [hunk 1]
```diff
@@ -41,20 +41,33 @@ func NewPredicates(fn PredicateFn) predicate.Funcs {
...
 
-// TODO(negz): Use separate predicates for resource claims and managed
-// resources once controller-runtime supports this.
-// https://github.com/kubernetes-sigs/controller-runtime/issues/572
-
-// HasClassReferenceKinds accepts objects that reference the supplied
-// non-portable class kind, either directly or indirectly via a portable class
-// kind.
-func HasClassReferenceKinds(c client.Client, oc runtime.ObjectCreater, k ClassKinds) PredicateFn {
+// AnyOf accepts objects that pass any of the supplied the supplied predicate
```
Comment (hasheddan) [inline pkg/resource/predicates.go]: This is really cool! Wonder if any upstream k8s projects (likely controller-runtime) would want to add this as part of the `NewControllerManagedBy` chain.

Comment (jbw976) [inline pkg/resource/predicates.go]: nit: stutter?
> the supplied the supplied predicate

File: pkg/resource/predicates.go [hunk 2]
```diff
@@ -41,20 +41,33 @@ func NewPredicates(fn PredicateFn) predicate.Funcs {
...
+		if !ok {
+			return false
 		}
 
-		return HasIndirectClassReferenceKind(c, oc, k)(obj)
+		if r.GetResourceReference() == nil {
+			return false
+		}
+
+		return r.GetResourceReference().GroupVersionKind() == schema.GroupVersionKind(k)
```
Comment (jbw976) [inline pkg/resource/predicates.go]: i was wondering about this the other day, do we need to do any `ToLower()` normalization or `EqualFold()` comparison in these predicates?

Comment (negz) [reply pkg/resource/predicates.go]: The type metadata supplied to the predicate will always match the actual case of our API types (e.g. `RDSInstanceClass`, not `rdsinstanceclass`), so the question is whether we want to require our users to match the actual case. I'm ambivalent about that so I'll avoid changing it for the time being. Everything will work fine as long as the correct case is used. It seems typical to do this in Kubernetes, i.e. I usually see `kind: Deployment`, not `kind: deployment`.

File: pkg/resource/predicates.go (hunk 1 above)
Comment (negz) [reply pkg/resource/predicates.go]: I believe they'd need to also be interested in taking on the `NewPredicates` pattern for it to work. I'm on the fence as to whether it's worth contributing; it's a fairly simple pattern.

**Pull request merged**